storage/backup/backup-cli/src/backup_types/transaction/restore.rs

```diff
@@ -422,27 +441,52 @@ impl TransactionRestoreBatchController {
     async fn replay_transactions(
         &self,
         restore_handler: &RestoreHandler,
-        txns_to_execute_stream: impl Stream<Item = Result<(Transaction, TransactionInfo)>>,
+        txns_to_execute_stream: impl Stream<
+            Item = Result<(Transaction, TransactionInfo, WriteSet, Vec<ContractEvent>)>,
+        >,
     ) -> Result<()> {
         let first_version = self.replay_from_version.unwrap();
         restore_handler.reset_state_store();
         let replay_start = Instant::now();
         let db = DbReaderWriter::from_arc(Arc::clone(&restore_handler.aptosdb));
         let chunk_replayer = Arc::new(ChunkExecutor::<AptosVM>::new(db));
+        let begin_version = std::cmp::max(
+            first_version,
+            restore_handler.get_next_expected_transaction_version()?,
+        );
 
         let db_commit_stream = txns_to_execute_stream
             .try_chunks(BATCH_SIZE)
             .err_into::<anyhow::Error>()
-            .map_ok(|chunk| {
-                let (txns, txn_infos): (Vec<_>, Vec<_>) = chunk.into_iter().unzip();
+            .enumerate()
+            .map(|(idx, chunk_res)| chunk_res.map(|chunk| (idx, chunk)))
+            .map_ok(|(idx, chunk)| {
+                let chunk_len = chunk.len();
+                let (txns, txn_infos, write_sets, events): (Vec<_>, Vec<_>, Vec<_>, Vec<_>) =
+                    chunk.into_iter().multiunzip();
                 let chunk_replayer = chunk_replayer.clone();
+                let chunk_begin_version = chunk_replayer.current_committed_version().unwrap();
```

Comment:
the problem is the version you get here is not the same as when the async closure is actually ran (later).

Reply:
Proly move the logic into the closure, or let chunk_replayer take the entire lookup table.